indy_common/authorizer.py

```diff
@@ -0,0 +1,147 @@
+        try:
+            owner = self.cache.getOwnerFor(request.operation[TARGET_NYM], isCommitted=False)
+        except KeyError:
+            """Not found in idrCache"""
+            owner = None
+        return owner
+
+    def _get_req_owner(self, request: Request):
+        return request.identifier
+
+    def _get_txn_owner(self, request: Request):
+        txn_type = request.operation[TXN_TYPE]
+        if txn_type == NYM:
+            return self._get_txn_owner_for_nym(request)
+        return None
+
+    def is_owner_accepted(self, request: Request, constraint: AuthConstraint):
+        if constraint.need_to_be_owner and \
+                self._get_req_owner(request) != self._get_txn_owner(request):
+            return False
+        return True
+
+    def is_role_accepted(self, request: Request, auth_constraint: AuthConstraint):
+        role = self.get_role(request)
+        if role:
+            return role == auth_constraint.role or auth_constraint.role == '*'
+
+    def is_sig_count_accepted(self, request: Request, auth_constraint: AuthConstraint):
+        sig_count = 1
+        if auth_constraint.sig_count != 1:
+            sig_count = self.get_sig_count(request)
+
+        return sig_count >= auth_constraint.sig_count
+
+    def authorize(self, request: Request, auth_constraint: AuthConstraint, auth_action: AbstractAuthAction=None):
+        if not self.is_role_accepted(request, auth_constraint):
+            return False, "role is not accepted"
+        if not self.is_sig_count_accepted(request, auth_constraint):
+            return False, "count of signatures is not accepted"
+        if not self.is_owner_accepted(request, auth_constraint):
+            return False, "actor must be owner"
+        return True, ""
+
+
+class CompositeAuthorizer(AbstractAuthorizer):
+    def __init__(self):
+        super().__init__()
+        self.authorizers = {}
+
+    def register_authorizer(self, authorizer, auth_constraint_id=ROLE_CONSTRAINT_ID):
+        authorizer.parent = self
+        self.authorizers.setdefault(auth_constraint_id, []).append(authorizer)
+
+    def authorize(self, request: Request, auth_constraint: AuthConstraint, auth_action: AbstractAuthAction):
+        for authorizer in self.authorizers.get(auth_constraint.constraint_id):
+            authorized, reason = authorizer.authorize(request=request,
+                                                      auth_constraint=auth_constraint,
+                                                      auth_action=auth_action)
+            if not authorized:
+                raise AuthValidationError("Validation error with reason: {}".format(reason))
+            return True, ""
+
+
+class AndAuthorizer(CompositeAuthorizer):
+
+    def authorize(self, request, auth_constraint: AuthConstraintAnd, auth_action: AbstractAuthAction):
+        for constraint in auth_constraint.auth_constraints:
+            authorized, reason = self.parent.authorize(request=request,
```

Comment:
Why do we call `parent` here? Should we just call `super().authorize`?

Reply:
in case of instantiating from class with Composite as a parent, attribute `authorizers` will be empty and super().authorize will not continue iterating over other authorizers. Therefore, we need to have a reference to object with registered authotizers. Like in Composite pattern

Reply:
But this should call Composite's authorize, not the actual parent authorize. If we have And/Or owned/parent by other And/Ors, then it may lead to calling an incorrect parent (And/Or instead of Composte's authorize).